File: fraiseql-go/examples/basic_schema.go
package main

import (
	"log"

	"github.com/fraiseql/fraiseql-go/fraiseql"
)

// User type representing a user in the system
type User struct {
	ID        int    `fraiseql:"id,type=Int"`
	Name      string `fraiseql:"name,type=String"`
	Email     string `fraiseql:"email,type=String"`
	CreatedAt string `fraiseql:"createdAt,type=String"`
	IsActive  bool   `fraiseql:"isActive,type=Boolean"`
}

// Post type representing a blog post
type Post struct {
	ID        int    `fraiseql:"id,type=Int"`
	Title     string `fraiseql:"title,type=String"`
	Content   string `fraiseql:"content,type=String"`
	AuthorID  int    `fraiseql:"authorId,type=Int"`
	Published bool   `fraiseql:"published,type=Boolean"`
	CreatedAt string `fraiseql:"createdAt,type=String"`
}

// Initialize queries and mutations using init() function
func init() {
	// Query: Get all users with pagination
	fraiseql.NewQuery("users").
		ReturnType(User{}).
		ReturnsArray(true).
		Config(map[string]interface{}{
			"sql_source": "v_user",
			"auto_params": map[string]bool{
				"limit":     true,
				"offset":    true,
				"where":     true,
				"order_by":  true,
			},
		}).
		Arg("limit", "Int", 10).
		Arg("offset", "Int", 0).
		Arg("isActive", "Boolean", nil, true).
		Description("Get list of users with pagination").
		Register()

	// Query: Get a single user by ID
	fraiseql.NewQuery("user").
		ReturnType(User{}).
		Config(map[string]interface{}{
			"sql_source": "v_user",
		}).
		Arg("id", "Int", nil).
		Description("Get a single user by ID").
		Register()

	// Query: Get posts with optional filtering
	fraiseql.NewQuery("posts").
		ReturnType(Post{}).
		ReturnsArray(true).
		Config(map[string]interface{}{
			"sql_source": "v_post",
			"auto_params": map[string]bool{
				"limit":     true,
				"offset":    true,
				"where":     true,
				"order_by":  true,
			},
		}).
		Arg("authorId", "Int", nil, true).
		Arg("published", "Boolean", true).
		Arg("limit", "Int", 10).
		Arg("offset", "Int", 0).
		Description("Get list of posts with filtering").
		Register()

	// Mutation: Create a new user
	fraiseql.NewMutation("createUser").
		ReturnType(User{}).
		Config(map[string]interface{}{
			"sql_source": "fn_create_user",
			"operation":  "CREATE",
		}).
		Arg("name", "String", nil).
		Arg("email", "String", nil).
		Description("Create a new user").
		Register()

	// Mutation: Update an existing user
	fraiseql.NewMutation("updateUser").
		ReturnType(User{}).
		Config(map[string]interface{}{
			"sql_source": "fn_update_user",
			"operation":  "UPDATE",
		}).
		Arg("id", "Int", nil).
		Arg("name", "String", nil, true).
		Arg("email", "String", nil, true).
		Arg("isActive", "Boolean", nil, true).
		Description("Update an existing user").
		Register()

	// Mutation: Delete a user
	fraiseql.NewMutation("deleteUser").
		ReturnType(User{}).
		Config(map[string]interface{}{
			"sql_source": "fn_delete_user",
			"operation":  "DELETE",
		}).
		Arg("id", "Int", nil).
		Description("Delete a user").
		Register()

	// Mutation: Create a new blog post
	fraiseql.NewMutation("createPost").
		ReturnType(Post{}).
		Config(map[string]interface{}{
			"sql_source": "fn_create_post",
			"operation":  "CREATE",
		}).
		Arg("title", "String", nil).
		Arg("content", "String", nil).
		Arg("authorId", "Int", nil).
		Description("Create a new blog post").
		Register()

	// Mutation: Publish a post
	fraiseql.NewMutation("publishPost").
		ReturnType(Post{}).
		Config(map[string]interface{}{
			"sql_source": "fn_publish_post",
			"operation":  "UPDATE",
		}).
		Arg("id", "Int", nil).
		Description("Publish a blog post").
		Register()
}

func main() {
	// Register the types
	if err := fraiseql.RegisterTypes(User{}, Post{}); err != nil {
		log.Fatal(err)
	}

	// Export schema to JSON
	if err := fraiseql.ExportSchema("schema.json"); err != nil {
		log.Fatal(err)
	}

	log.Println("\n✅ Schema exported successfully!")
	log.Println("   Next steps:")
	log.Println("   1. Compile schema: fraiseql-cli compile schema.json")
	log.Println("   2. Start server: fraiseql-server --schema schema.compiled.json")
}

File: sdks/official/fraiseql-java/src/main/java/com/fraiseql/examples/BasicSchema.java
package com.fraiseql.examples;

import com.fraiseql.core.*;

/**
 * BasicSchema - Complete example of FraiseQL Java authoring layer
 *
 * This example demonstrates:
 * 1. Defining GraphQL types with @GraphQLType and @GraphQLField
 * 2. Registering types with the schema
 * 3. Creating queries with the fluent API
 * 4. Creating mutations with the fluent API
 * 5. Exporting the schema to schema.json
 *
 * Output: schema.json - Ready for fraiseql-cli compile
 *
 * Usage:
 *   mvn exec:java -Dexec.mainClass="com.fraiseql.examples.BasicSchema"
 */
public class BasicSchema {

    public static void main(String[] args) {
        try {
            System.out.println("FraiseQL Java - BasicSchema Example");
            System.out.println("====================================\n");

            // Register types
            System.out.println("1. Registering GraphQL types...");
            FraiseQL.registerTypes(User.class, Post.class, Comment.class);
            System.out.println("   ✓ User, Post, Comment registered\n");

            // Register queries
            System.out.println("2. Registering GraphQL queries...");
            FraiseQL.query("users")
                .returnType(User.class)
                .returnsArray(true)
                .arg("limit", "Int")
                .arg("offset", "Int")
                .description("Get all users with pagination")
                .register();
            System.out.println("   ✓ users query registered");

            FraiseQL.query("user")
                .returnType(User.class)
                .arg("id", "Int")
                .description("Get a user by ID")
                .register();
            System.out.println("   ✓ user query registered");

            FraiseQL.query("posts")
                .returnType(Post.class)
                .returnsArray(true)
                .arg("userId", "Int")
                .description("Get posts for a specific user")
                .register();
            System.out.println("   ✓ posts query registered");

            FraiseQL.query("post")
                .returnType(Post.class)
                .arg("id", "Int")
                .description("Get a specific post")
                .register();
            System.out.println("   ✓ post query registered");

            FraiseQL.query("comments")
                .returnType(Comment.class)
                .returnsArray(true)
                .arg("postId", "Int")
                .description("Get comments for a post")
                .register();
            System.out.println("   ✓ comments query registered\n");

            // Register mutations
            System.out.println("3. Registering GraphQL mutations...");
            FraiseQL.mutation("createUser")
                .returnType(User.class)
                .arg("name", "String")
                .arg("email", "String")
                .description("Create a new user")
                .register();
            System.out.println("   ✓ createUser mutation registered");

            FraiseQL.mutation("updateUser")
                .returnType(User.class)
                .arg("id", "Int")
                .arg("name", "String")
                .arg("email", "String")
                .description("Update an existing user")
                .register();
            System.out.println("   ✓ updateUser mutation registered");

            FraiseQL.mutation("deleteUser")
                .returnType(User.class)
                .arg("id", "Int")
                .description("Delete a user")
                .register();
            System.out.println("   ✓ deleteUser mutation registered");

            FraiseQL.mutation("createPost")
                .returnType(Post.class)
                .arg("userId", "Int")
                .arg("title", "String")
                .arg("content", "String")
                .description("Create a new post")
                .register();
            System.out.println("   ✓ createPost mutation registered");

            FraiseQL.mutation("createComment")
                .returnType(Comment.class)
                .arg("postId", "Int")
                .arg("userId", "Int")
                .arg("text", "String")
                .description("Create a comment on a post")
                .register();
            System.out.println("   ✓ createComment mutation registered\n");

            // Export schema
            System.out.println("4. Exporting schema to schema.json...");
            String outputPath = "schema.json";
            FraiseQL.exportSchema(outputPath);
            System.out.println("   ✓ Schema exported to " + outputPath + "\n");

            // Print summary
            SchemaRegistry registry = FraiseQL.getRegistry();
            System.out.println("Schema Summary:");
            System.out.println("--------------");
            System.out.println("Types:     " + registry.getAllTypes().size());
            System.out.println("Queries:   " + registry.getAllQueries().size());
            System.out.println("Mutations: " + registry.getAllMutations().size());
            System.out.println("\nNext steps:");
            System.out.println("1. Run: fraiseql-cli compile schema.json -o schema.compiled.json");
            System.out.println("2. Run: fraiseql-server --schema schema.compiled.json --port 8000");
            System.out.println("3. Test: curl -X POST http://localhost:8000/graphql ...");

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    /**
     * User type with id, name, email, and createdAt
     */
    @GraphQLType(description = "A user account")
    public static class User {
        @GraphQLField(description = "User ID")
        public int id;

        @GraphQLField(description = "User's name")
        public String name;

        @GraphQLField(description = "User's email address")
        public String email;

        @GraphQLField(name = "created_at", description = "When the user was created")
        public String createdAt;
    }

    /**
     * Post type with id, userId, title, content, and createdAt
     */
    @GraphQLType(description = "A blog post")
    public static class Post {
        @GraphQLField(description = "Post ID")
        public int id;

        @GraphQLField(description = "ID of the user who created this post")
        public int userId;

        @GraphQLField(description = "Post title")
        public String title;

        @GraphQLField(description = "Post content")
        public String content;

        @GraphQLField(name = "created_at", description = "When the post was created")
        public String createdAt;
    }

    /**
     * Comment type with id, postId, userId, text, and createdAt
     */
    @GraphQLType(description = "A comment on a post")
    public static class Comment {
        @GraphQLField(description = "Comment ID")
        public int id;

        @GraphQLField(description = "ID of the post this comment is on")
        public int postId;

        @GraphQLField(description = "ID of the user who created this comment")
        public int userId;

        @GraphQLField(description = "Comment text")
        public String text;

        @GraphQLField(name = "created_at", description = "When the comment was created")
        public String createdAt;
    }
}
